### DISCUSSION TITLE
validateAndCreate fails when loading property that is DataCollection?

### DISCUSSION THREAD

**Discussion by wivaku** [spatie/laravel-data#417] [Q&A]
Why does validation fail for this one?

```
class SongData extends Data
{
    public function __construct(
      // ...
    ) {
    }
}

class AlbumData extends Data
{
    public function __construct(
        // ...

        #[DataCollectionOf(SongData::class)]
        public DataCollection $songs,
    ) {
    }
}

// fails: the songs field must be an array  <===
AlbumData::validateAndCreate([
  // ...
  "songs" => SongData::collection($songs),
]);

// works
AlbumData::validateAndCreate([
  // ...
  "songs" => $songs,
]);

// works
AlbumData::from([
  // ...
  "songs" => SongData::collection($songs), // as collection
  // "songs" => $songs, // or as array
]);
```

So: `from()` works with array and DataCollection, but `validateAndCreate()` only works with array.

`getValidationRules` indeed shows:
```
  "songs" => array:2 [
    0 => "present"
    1 => "array"
  ]
```

But that logic / the flow is unclear to me. I specify `$songs` will be a `DataCollection`, I would then expect that the validator knows the `DataCollection` has an array "inside".

In general I find my self confused often about the difference between `validateAndCreate` and `from`.  
I usually do want to validate the data (from array, not request), but *after* mapping/casting/etc, not *before*.

BTW it would be really helpful to bring the documentation pages more together / in sync. E.g. the [quick guide](https://spatie.be/docs/laravel-data/v3/getting-started/quickstart#content-nesting-data-objects-and-collections) mentions passing a regular array as nested collection using `from`, but the detailed [collections page](https://spatie.be/docs/laravel-data/v3/as-a-data-transfer-object/collections) does not.

**Comment by rubenvanassche** [comment]
From won't do validation by default, validateAndCreate will which means the DataCollection is passed to the Laravel validator which doesn't know a DataCollection is an array basically. In v4 we'll fase out DataCollections in favor for arrays and collections which should fix this problem.

> In general I find my self confused often about the difference between validateAndCreate and from.
> I usually do want to validate the data (from array, not request), but after mapping/casting/etc, not before.

You can do this by changing the validation Pipe to always validate everything passed to it.

>BTW it would be really helpful to bring the documentation pages more together / in sync. E.g. the [quick guide](https://spatie.be/docs/laravel-data/v3/getting-started/quickstart#content-nesting-data-objects-and-collections) mentions passing a regular array as nested collection using from, but the detailed [collections page](https://spatie.be/docs/laravel-data/v3/as-a-data-transfer-object/collections) does not.

The quick start guide needs an update, anyone is welcome PR'ing one. I don't think we need to mention that you can pas an array nested, these things should just work. But you're always welcome to send in a PR updating the docs.
